gpcheckperf: add buffer size parameter

red1452 opened the pull request:
Use '--buffer-size' flag with size in kilobytes to set buffer size,
which will be used at gpnetbenchClient.
It is optional parameter. Default value is 8 kilobytes.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

jnihal commented on gpMgmt/test/behave/mgmt_utils/gpcheckperf.feature:
```diff
@@ -52,3 +52,16 @@ Feature: Tests for gpcheckperf
+    And   gpcheckperf should not print "NOTICE: -t is deprecated " to stdout
+
+  @concourse_cluster
+  Scenario: gpcheckperf runs sequential network test with buffer size flag
```
Can we update the name of this scenario to represent the actual scenario it is checking? Right now it is too generic.

red1452 replied:
Done

jnihal commented on gpMgmt/bin/gpcheckperf:
```diff
@@ -275,6 +278,12 @@ def parseCommandLine():
         print('[INFO] Invalid network duration specified.  Using default (15 seconds)')
 
+    if GV.opt['--buffer-size'] > 0 and GV.opt['--netclient'].find('netperf') >= 0:
+        print '[Warning] Applying the --buffer-size option is not possible when the --netperf option is enabled.'
```
Use parentheses around `print` statements. `python3` does not support this format.

red1452 replied:
Fixed.

jnihal commented on gpMgmt/bin/gpcheckperf:
```diff
@@ -226,6 +227,8 @@ def parseCommandLine():
             GV.opt['--netserver'] = 'netserver'
             GV.opt['--netclient'] = 'netperf'
+        elif switch == '--buffer-size':
+            GV.opt[switch] = val
```
We need to convert the data type value to an `integer` here.

red1452 replied:
Fixed.

red1452 commented on gpMgmt/test/behave/mgmt_utils/gpcheckperf.feature:
```diff
@@ -73,3 +73,16 @@ Feature: Tests for gpcheckperf
+  @concourse_cluster
+  Scenario: gpcheckperf runs sequential network test with buffer size flag
+    Given the database is running
+    When  the user runs "gpcheckperf -h mdw -h sdw1 -d /data/gpdata/ -r n --buffer-size=8"
```
Fixed.

red1452 commented on gpMgmt/test/behave/mgmt_utils/gpcheckperf.feature:
```diff
@@ -73,3 +73,16 @@ Feature: Tests for gpcheckperf
+  @concourse_cluster
+  Scenario: gpcheckperf runs sequential network test with buffer size flag and netperf option
+    Given the database is running
+    When  the user runs "gpcheckperf -h mdw -h sdw1 -d /data/gpdata/ -r n --buffer-size=8 --netperf"
```
Fixed.

red1452 commented on gpMgmt/bin/gpcheckperf:
```diff
@@ -275,6 +278,12 @@ def parseCommandLine():
+        print('[Warning] Applying the --buffer-size option is not possible when the --netperf option is enabled.')
+
+    if GV.opt['--buffer-size'] <= 0:
+        GV.opt['--buffer-size'] = 8;
```
Fixed.

jnihal commented:
Thanks for addressing the review comments. Could you rebase this with the latest GPDB code? Want to run a pipeline to make sure everything is good.

red1452 commented:
> Thanks for addressing the review comments. Could you rebase this with the latest GPDB code? Want to run a pipeline to make sure everything is good.

I updated branch and rebased (squashed) commits.

piyushc01 commented on gpMgmt/bin/gpcheckperf:
```diff
@@ -275,6 +278,12 @@ def parseCommandLine():
+        print('[Warning] Applying the --buffer-size option is not possible when the --netperf option is enabled.')
+
+    if GV.opt['--buffer-size'] <= 0:
+        GV.opt['--buffer-size'] = 8
```
gpbenchClient has a default buffer value of 32kb. 
Do we want to be in sync?
```
gpnetbenchClient -h
usage: gpnetbenchClient -p PORT -H HOST [OPTIONS]

 -p PORT        port to connect to for the server
 -H HOST        hostname to connect to for the server
 -l SECONDS     number of seconds to sample the network, default is 60
 -P {0|1}       0 (don't) or 1 (do) display headers in the output, default is 1
 -b SIZE        size of the send buffer in kilobytes, default is 32
 -h             show this help message
```

red1452 replied:
This patch was done because of problem with gpcheckperf at network with mtu=9000:
```
[Warning] connection between kdv-adb6201s1 and kdv-adb6201s2 is no good
```
Buffer with size 32K is too big and leads to use fragmentation at network connection. Fragmentation leads to a decrease in network bandwidth.
If by default you use network with more mtu, I can change default value of buffer size.

piyushc01 replied:
Sounds good. Thanks

piyushc01 reviewed (approved):
Looks good to me

shrakesh commented on gpMgmt/test/behave/mgmt_utils/gpcheckperf.feature:
```diff
@@ -73,3 +73,16 @@ Feature: Tests for gpcheckperf
+  Scenario: gpcheckperf runs sequential network test with buffer size flag and netperf option
+    Given the database is running
+    When  the user runs "gpcheckperf -h cdw -h sdw1 -d /data/gpdata/ -r n --buffer-size=8 --netperf"
+    Then  gpcheckperf should print "Applying the --buffer-size option is not possible when the --netperf option is enabled." to stdout
```
we should add a test for the default value case,  to show if the option is not provided buffer-size value is set to default. 
if possible we should try to assert the cmd string created from the following code base. 
```
 cmd = ('%s -H %s -p %d -l %s -P 0 -b %s'
               % (netperf_path, y, netserver_port, sec, GV.opt['--buffer-size']))
```

shrakesh reviewed (commented):
Added a few suggestions to look on:
1. messages and help text 
2. restructure the code.
3. few test scenarios to add.

nimish350 commented on gpMgmt/bin/gpcheckperf:
```diff
@@ -275,6 +278,12 @@ def parseCommandLine():
         print('[INFO] Invalid network duration specified.  Using default (15 seconds)')
 
+    if GV.opt['--buffer-size'] > 0 and GV.opt['--netclient'].find('netperf') >= 0:
+        print('[Warning] Applying the --buffer-size option is not possible when the --netperf option is enabled.')
```
"ignored" sounds good. +1

nimish350 commented on gpMgmt/bin/gpcheckperf:
```diff
@@ -29,6 +29,7 @@ Usage: gpcheckperf <options>
     -f file    : a file listing all hosts to connect to
     --duration : how long to run network test (default 5 seconds)
     --netperf  : use netperf instead of gpnetbenchServer/gpnetbenchClient
+    --buffer-size=size : size of the send buffer in kilobytes, which is used by gpnetbenchClient
```
I would go with
--buffer-size : the size of the send buffer in kilobytes ( default 8 kilobytes)

nimish350 commented on gpMgmt/bin/gpcheckperf:
```diff
@@ -275,6 +278,12 @@ def parseCommandLine():
         GV.opt['--duration'] = 15
         print('[INFO] Invalid network duration specified.  Using default (15 seconds)')
 
+    if GV.opt['--buffer-size'] > 0 and GV.opt['--netclient'].find('netperf') >= 0:
```
Can we change this to
[INFO] --buffer-size value not specified or is invalid. Using default (8 kilobytes)

red1452 replied:
I think, that `[INFO] --buffer-size value is not specified or invalid. Using default (8 kilobytes)` is more correct.

shrakesh commented on gpMgmt/test/behave/mgmt_utils/gpcheckperf.feature:
```diff
@@ -73,3 +73,25 @@ Feature: Tests for gpcheckperf
+    Given the database is running
+    When  the user runs "gpcheckperf -h cdw -h sdw1 -d /data/gpdata/ -r n"
+    Then  gpcheckperf should return a return code of 0
+    And   gpcheckperf should print "--buffer-size value is not specified or invalid. Using default (8 kilobytes)" to stdout
```
the test is failing: 
Exception: Expected stdout string '--buffer-size value is not specified or invalid. Using default (8 kilobytes)' and found: '[INFO] --buffer-size value is not specified or invalid. Using default (8 kilobytes)

red1452 replied:
Fixed. I missed escaping the brackets.

shrakesh commented on gpMgmt/bin/gpcheckperf:
```diff
@@ -275,6 +278,17 @@ def parseCommandLine():
+            print('[INFO] --buffer-size value is not specified or invalid. Using default (8 kilobytes)')
+            GV.opt['--buffer-size'] = 8
+
+    if GV.opt['--buffer-size'] <= 0:
```
@red1452 do we really require the following section now, I think we can remove line no 289 and 290 : 
```
if GV.opt['--buffer-size'] <= 0:
      GV.opt['--buffer-size'] = 8
```
if I am understanding it correctly, if --netclient is not enabled in that case if the buffer-size is negative or 0 it will be setting it to default(8 kbytes) with the info message. 
please correct me if I am getting it differently.

red1452 replied:
You are right. Removed.

red1452 commented on gpMgmt/bin/gpcheckperf:
```diff
@@ -540,8 +554,11 @@ def spawnNetperfTestBetween(x, y, netperf_path, netserver_port, sec=5):
+        cmd = ('%s -H %s -p %d -l %s -P 0 -b %s'
+               % (netperf_path, y, netserver_port, sec, GV.opt['--buffer-size']))
+
+    assert cmd is not None
```
I added line `And   gpcheckperf should print "gpnetbenchClient -H cdw -p 23000 -l 15 -P 0 -b 8" to stdout` to behave test. This line is printed if we use `-v` flag.

shrakesh reviewed (approved):
Looking good to me.
